File: android/data/src/main/kotlin/kg/ayant/app/data/firestore/FirestoreSchema.kt
package kg.ayant.app.data.firestore

/**
 * **Единственное место, где записаны имена коллекций и полей Firestore.**
 *
 * Раньше каждое имя было строковым литералом и встречалось 2–3 раза: в чтении
 * (`getString("venueName")`), в записи (`"venueName" to …`), иногда ещё в запросе.
 * Переименование поля требовало найти их все — а промах не ломал сборку, он тихо
 * возвращал `null` и подставлял дефолт. Теперь имя объявлено один раз, и опечатка
 * в нём — ошибка компиляции.
 *
 * Важно: одни и те же документы пишут ещё три поверхности — iOS-приложение,
 * Cloud Functions (`functions/src/index.ts`) и админ-панель (`docs/admin/index.html`).
 * Компилятор их не проверяет, поэтому при переименовании поля правьте все четыре.
 *
 * Зеркалит `SAN/Firebase/FirestoreSchema.swift` на iOS 1:1.
 */
object FS {

    // ── Коллекции ────────────────────────────────────────────────────────────

    object Collection {
        const val VENUES = "venues"
        const val DEALS = "deals"
        const val REVIEWS = "reviews"
        const val COUPONS = "coupons"
        const val GIFT_COUPONS = "giftCoupons"
        const val LOYALTY_CARDS = "loyaltyCards"
        const val VENUE_POINTS = "venuePoints"
        const val BONUS_GRANTS = "bonusGrants"
        const val REDEMPTIONS = "redemptions"
        const val REFERRALS = "referrals"
        const val RANKING_EVENTS = "rankingEvents"
        const val ANALYTICS = "analytics"
        const val ANALYTICS_EVENTS = "analyticsEvents"
        const val USER_TOKENS = "userTokens"
        const val PUSH_CAMPAIGNS = "pushCampaigns"
        const val CONFIG = "config"
        /** Подколлекция `analytics/{venueID}/days/{yyyy-MM-dd}`. */
        const val DAYS = "days"
    }

    /** Документы с фиксированным id. */
    object Document {
        const val RANKING_WEIGHTS = "rankingWeights"
    }

    // ── venues/{id} ──────────────────────────────────────────────────────────

    object VenueDoc {
        const val NAME = "name"
        const val CATEGORY = "category"
        const val DISTRICT = "district"
        const val ADDRESS = "address"
        const val PHONE = "phone"
        const val EMOJI = "emoji"
        const val IMAGE_URL = "imageURL"
        const val RATING = "rating"
        const val REVIEW_COUNT = "reviewCount"
        const val IS_VERIFIED = "isVerified"
        const val SAVED_BY_COUNT = "savedByCount"
        /** Слаг города. Внимание: поле называется `city`, а в модели — `citySlug`. */
        const val CITY = "city"
        const val LATITUDE = "latitude"
        const val LONGITUDE = "longitude"
        const val TODAY_SPECIAL = "todaySpecial"
        const val OPEN_HOUR = "openHour"
        const val CLOSE_HOUR = "closeHour"
        const val WEEK_HOURS = "weekHours"
        const val PDF_MENU_URL = "pdfMenuURL"
        const val PHOTO_EMOJIS = "photoEmojis"
        const val OWNER_ID = "ownerID"
        const val STATUS = "status"
        const val IS_PAUSED = "isPaused"
        const val WHATSAPP = "whatsapp"
        const val INSTAGRAM = "instagram"
        const val TELEGRAM = "telegram"
        const val BRANCHES = "branches"
        const val BOOSTED_UNTIL = "boostedUntil"
        // Карта лояльности.
        const val LOYALTY_ENABLED = "loyaltyEnabled"
        const val LOYALTY_GOAL = "loyaltyGoal"
        const val LOYALTY_REWARD = "loyaltyReward"
        const val COUPONS_ENABLED = "couponsEnabled"
        // Баллы САН — эти поля ведёт админ-панель.
        const val POINTS_ENABLED = "pointsEnabled"
        const val POINTS_MODE = "pointsMode"
        const val POINTS_FLAT = "pointsFlat"
        const val POINTS_BANDS = "pointsBands"
        const val CASHBACK_PERCENT = "cashbackPercent"
        const val POINTS_REWARDS = "pointsRewards"
        const val POINTS_EXPIRY_MONTHS = "pointsExpiryMonths"
        const val REDEEM_MODE = "redeemMode"
        const val EARN_COOLDOWN_MINUTES = "earnCooldownMinutes"
    }

    // ── deals/{id} ───────────────────────────────────────────────────────────

    object DealDoc {
        const val VENUE_ID = "venueID"
        const val TYPE = "type"
        const val TITLE = "title"
        const val DETAILS = "details"
        /** Условия акции: массив строк, по строке на пункт. */
        const val TERMS = "terms"
        const val EMOJI = "emoji"
        const val OLD_PRICE = "oldPrice"
        const val NEW_PRICE = "newPrice"
        const val DISCOUNT_PERCENT = "discountPercent"
        const val VALID_UNTIL = "validUntil"
        const val STATUS = "status"
        const val START_DATE = "startDate"
        const val IMAGE_URL = "imageURL"
        const val IMAGE_URLS = "imageURLs"
        /** Ключ партиционирования по городам — то же поле `city`, что у заведения. */
        const val CITY = "city"
    }

    // ── reviews/{id} ─────────────────────────────────────────────────────────

    object ReviewDoc {
        const val VENUE_ID = "venueID"
        const val AUTHOR_ID = "authorID"
        const val AUTHOR_NAME = "authorName"
        const val RATING = "rating"
        const val TEXT = "text"
        const val PHOTOS = "photos"
        const val CREATED_AT = "createdAt"
        const val UPDATED_AT = "updatedAt"
        const val HOST_REPLY = "hostReply"
        const val ITEM_ID = "itemID"
        const val ITEM_NAME = "itemName"
        const val VERIFIED_VISIT = "verifiedVisit"
        /** Ключ партиционирования по городам — то же поле `city`, что у заведения. */
        const val CITY = "city"
    }

    /** Вложенная карта `reviews/{id}.hostReply`. */
    object HostReplyField {
        const val TEXT = "text"
        const val CREATED_AT = "createdAt"
        const val UPDATED_AT = "updatedAt"
    }

    // ── Вложенные объекты внутри venues/{id} ─────────────────────────────────

    /** Элемент массива `branches` (филиал). */
    object BranchField {
        const val ID = "id"
        const val ADDRESS = "address"
        const val LATITUDE = "latitude"
        const val LONGITUDE = "longitude"
        const val PHONE = "phone"
    }

    /** Элемент массива `weekHours` (часы одного дня). */
    object DayHoursField {
        const val CLOSED = "closed"
        const val OPEN = "open"
        const val CLOSE = "close"
    }

    /** Элемент массива `pointsBands` (диапазон чек→баллы). */
    object PointsBandField {
        const val MAX_AMOUNT = "maxAmount"
        const val POINTS = "points"
    }

    /** Элемент массива `pointsRewards` (награда за баллы). */
    object PointsRewardField {
        const val ID = "id"
        const val TYPE = "type"
        const val TITLE = "title"
        const val COST = "cost"
        const val RATIO = "ratio"
        const val ACTIVE = "active"
    }

    // ── Кошельки гостя ───────────────────────────────────────────────────────

    object CouponDoc {
        const val CODE = "code"
        const val USER_ID = "userID"
        const val VENUE_ID = "venueID"
        const val VENUE_NAME = "venueName"
        const val TITLE = "title"
        const val KIND = "kind"
        const val DEAL_ID = "dealID"
        const val USED = "used"
        const val CREATED_AT = "createdAt"
    }

    object LoyaltyCardDoc {
        const val USER_ID = "userID"
        const val VENUE_ID = "venueID"
        const val VENUE_NAME = "venueName"
        const val STAMPS = "stamps"
        const val COMPLETED_ROUNDS = "completedRounds"
        const val GOAL = "goal"
        const val REWARD = "reward"
    }

    object VenuePointsDoc {
        const val USER_ID = "userID"
        const val VENUE_ID = "venueID"
        const val VENUE_NAME = "venueName"
        const val BALANCE = "balance"
        const val LIFETIME_EARNED = "lifetimeEarned"
        const val LIFETIME_REDEEMED = "lifetimeRedeemed"
    }

    object BonusGrantDoc {
        const val USER_ID = "userID"
        const val AMOUNT = "amount"
        const val CLAIMED = "claimed"
    }

    /** `redemptions/{userID}_{dealID}` — детерминированный id, повторов не будет. */
    object RedemptionDoc {
        const val USER_ID = "userID"
        const val DEAL_ID = "dealID"
        const val VENUE_ID = "venueID"
        const val CREATED_AT = "createdAt"
        const val STATUS = "status"
        /** Значение поля `status` для свежей записи. */
        const val STATUS_NEW = "new"
    }

    /** `referrals/{inviteeID}` — один документ на приглашённого. */
    object ReferralDoc {
        const val INVITEE_ID = "inviteeID"
        const val REFERRER_ID = "referrerID"
        const val CREATED_AT = "createdAt"
    }

    /** `giftCoupons/{code}` — подарочный купон, забирается один раз. */
    object GiftCouponDoc {
        const val TITLE = "title"
        const val CODE = "code"
        const val FROM_NAME = "fromName"
        const val CLAIMED = "claimed"
        const val CLAIMED_AT = "claimedAt"
        const val CREATED_AT = "createdAt"
    }

    // ── Телеметрия ───────────────────────────────────────────────────────────

    /**
     * `analyticsEvents/{auto}` — клиентское событие; счётчики досчитывает
     * Cloud Function (писать в `analytics/…` напрямую клиенту запрещено).
     */
    object AnalyticsEventDoc {
        const val VENUE_ID = "venueID"
        const val METRIC = "metric"
        const val CREATED_AT = "createdAt"
    }

    /**
     * `analytics/{venueID}/days/{yyyy-MM-dd}` — счётчики за день.
     * Все поля, кроме `date`, суммируются как метрики.
     */
    object AnalyticsDayDoc {
        const val DATE = "date"
    }

    /**
     * `pushCampaigns/{auto}` — заявка на рассылку; админ одобряет в панели
     * (`status` → `approved`), дальше рассылает Cloud Function `sendPushCampaign`.
     *
     * `STATUS_PENDING` обязателен: админ-панель показывает к одобрению только
     * документы со `status == "pending"`, а функция шлёт только `"approved"`.
     * Любое другое значение = заявка навсегда застревает невидимой.
     */
    object PushCampaignDoc {
        const val HEADLINE = "headline"
        const val BODY = "body"
        const val CITY = "city"
        const val CATEGORY = "category"
        const val VENUE_ID = "venueID"
        const val DEAL_ID = "dealID"
        const val OWNER_ID = "ownerID"
        const val STATUS = "status"
        const val DELIVERED = "delivered"
        const val CREATED_AT = "createdAt"
        const val STATUS_PENDING = "pending"
    }

    /** `userTokens/{fcmToken}` — устройство для адресной рассылки. */
    object UserTokenDoc {
        const val CITY = "city"
        const val UID = "uid"
        const val UPDATED_AT = "updatedAt"
    }

    // ── Ответы Cloud Functions (scanCoupon / redeemVenuePoints) ──────────────
    //
    // Это не Firestore, а тело HTTP-ответа. Держим здесь по той же причине:
    // ключи заданы сервером в functions/src/index.ts и должны совпадать.

    object ScanResponse {
        const val OK = "ok"
        const val TITLE = "title"
        const val LOYALTY = "loyalty"
        const val STAMPS = "stamps"
        const val GOAL = "goal"
        const val REWARD_ISSUED = "rewardIssued"
        const val REWARD_TITLE = "rewardTitle"
        const val ERROR = "error"
        const val POINTS = "points"
        const val AWARDED = "awarded"
        const val BALANCE = "balance"
        // Поля запроса.
        const val CODE = "code"
        const val VENUE_ID = "venueID"
        const val BILL_AMOUNT = "billAmount"
        const val BAND_INDEX = "bandIndex"
        const val IDEMPOTENCY_KEY = "idempotencyKey"
        /** true — сервер распознал повтор по ключу и НЕ начислил снова. */
        const val REPLAYED = "replayed"
    }

    object RedeemResponse {
        const val OK = "ok"
        const val REDEEMED = "redeemed"
        const val BALANCE = "balance"
        const val REWARD_TITLE = "rewardTitle"
        const val SOM_OFF = "somOff"
        const val ERROR = "error"
        /** true — сервер распознал повтор по idempotencyKey и НЕ списал баллы снова. */
        const val REPLAYED = "replayed"
        const val IDEMPOTENCY_KEY = "idempotencyKey"
        // Поля запроса.
        const val VENUE_ID = "venueID"
        const val USER_ID = "userID"
        const val REWARD_ID = "rewardId"
        const val POINTS_TO_SPEND = "pointsToSpend"
    }

    /**
     * Data-payload push-уведомления (его собирает `sendPushCampaign`).
     * По этим ключам приложение решает, что открыть по тапу.
     */
    object PushPayload {
        const val DEAL_ID = "dealID"
        const val VENUE_ID = "venueID"
    }
}

/**
 * Слаги категорий/типов/статусов — тоже часть схемы: их пишет и админка.
 * Значения совпадают с `FSKeys` на iOS.
 */
object FSKeys {
    /** slug → отображаемое имя категории (RU). */
    val category = mapOf(
        "cafe" to "Кафе", "coffee" to "Кофейня", "fastfood" to "Фастфуд",
        "restaurant" to "Ресторан", "teahouse" to "Чайхана", "bakery" to "Пекарня",
    )

    const val DEAL_TYPE_DISCOUNT = "discount"
    const val DEAL_TYPE_PROMO = "promo"
    const val DEAL_TYPE_NOVELTY = "novelty"
    const val DEAL_TYPE_ANNOUNCEMENT = "announcement"

    const val DEAL_STATUS_ACTIVE = "active"
    const val DEAL_STATUS_PAUSED = "paused"
    const val DEAL_STATUS_EXPIRED = "expired"
    const val DEAL_STATUS_DRAFT = "draft"

    const val MODERATION_PENDING = "pending"
    const val MODERATION_APPROVED = "approved"
    const val MODERATION_REJECTED = "rejected"
}

File: AyantData/Sources/AyantData/FirestoreSchema.swift
import Foundation
import AyantDomain

/// **Единственное место, где записаны имена коллекций и полей Firestore.**
///
/// Раньше каждое имя было строковым литералом и встречалось 2–3 раза: в чтении
/// (`d["gradientFrom"]`), в записи (`"gradientFrom": …`), иногда ещё в запросе.
/// Переименование поля требовало найти их все — а промах не ломал сборку, он тихо
/// возвращал `nil` и подставлял дефолт. Теперь имя объявлено один раз, и опечатка
/// в нём — ошибка компиляции.
///
/// Важно: одни и те же документы пишут ещё две поверхности — Cloud Functions
/// (`functions/src/index.ts`) и админ-панель (`docs/admin/index.html`). Компилятор
/// их не проверяет, поэтому при переименовании поля правьте все четыре.
///
/// Зеркалит `data/firestore/FirestoreSchema.kt` на Android 1:1.
public enum FS {

    // MARK: - Коллекции

    public enum Collection {
        public static let venues = "venues"
        public static let deals = "deals"
        public static let reviews = "reviews"
        public static let hosts = "hosts"
        public static let categories = "categories"
        public static let coupons = "coupons"
        public static let giftCoupons = "giftCoupons"
        public static let loyaltyCards = "loyaltyCards"
        public static let venuePoints = "venuePoints"
        public static let bonusGrants = "bonusGrants"
        public static let redemptions = "redemptions"
        public static let referrals = "referrals"
        public static let rankingEvents = "rankingEvents"
        public static let analytics = "analytics"
        public static let analyticsEvents = "analyticsEvents"
        public static let pushCampaigns = "pushCampaigns"
        public static let userTokens = "userTokens"
        public static let config = "config"
        /// Подколлекция `analytics/{venueID}/days/{yyyy-MM-dd}`.
        public static let days = "days"
    }

    /// Документы с фиксированным id.
    public enum Document {
        public static let rankingWeights = "rankingWeights"
    }

    // MARK: - venues/{id}

    public enum VenueDoc {
        public static let name = "name"
        public static let category = "category"
        public static let district = "district"
        public static let address = "address"
        public static let phone = "phone"
        public static let emoji = "emoji"
        public static let gradientFrom = "gradientFrom"
        public static let gradientTo = "gradientTo"
        public static let imageURL = "imageURL"
        public static let rating = "rating"
        public static let reviewCount = "reviewCount"
        public static let isVerified = "isVerified"
        public static let savedByCount = "savedByCount"
        /// Слаг города. Внимание: поле называется `city`, а в модели — `citySlug`.
        public static let city = "city"
        public static let latitude = "latitude"
        public static let longitude = "longitude"
        public static let todaySpecial = "todaySpecial"
        public static let openHour = "openHour"
        public static let closeHour = "closeHour"
        public static let weekHours = "weekHours"
        public static let pdfMenuURL = "pdfMenuURL"
        public static let photoEmojis = "photoEmojis"
        public static let ownerID = "ownerID"
        public static let items = "items"
        public static let status = "status"
        public static let isPaused = "isPaused"
        public static let whatsapp = "whatsapp"
        public static let instagram = "instagram"
        public static let telegram = "telegram"
        public static let branches = "branches"
        public static let boostedUntil = "boostedUntil"
        // Карта лояльности.
        public static let loyaltyEnabled = "loyaltyEnabled"
        public static let loyaltyGoal = "loyaltyGoal"
        public static let loyaltyReward = "loyaltyReward"
        public static let couponsEnabled = "couponsEnabled"
        // Баллы САН. Эти поля пишет ТОЛЬКО админ-панель — хост их не трогает,
        // иначе сохранение из приложения затёрло бы настройки (см. HostVenueDTO).
        public static let pointsEnabled = "pointsEnabled"
        public static let pointsMode = "pointsMode"
        public static let pointsFlat = "pointsFlat"
        public static let pointsBands = "pointsBands"
        public static let cashbackPercent = "cashbackPercent"
        public static let pointsRewards = "pointsRewards"
        public static let pointsExpiryMonths = "pointsExpiryMonths"
        public static let redeemMode = "redeemMode"
        public static let earnCooldownMinutes = "earnCooldownMinutes"
    }

    // MARK: - deals/{id}

    public enum DealDoc {
        public static let venueID = "venueID"
        public static let type = "type"
        public static let title = "title"
        public static let details = "details"
        /// Условия акции: массив строк, по строке на пункт.
        public static let terms = "terms"
        public static let emoji = "emoji"
        public static let oldPrice = "oldPrice"
        public static let newPrice = "newPrice"
        public static let discountPercent = "discountPercent"
        public static let validUntil = "validUntil"
        public static let status = "status"
        public static let startDate = "startDate"
        public static let endDate = "endDate"
        public static let imageEmojis = "imageEmojis"
        public static let imageURL = "imageURL"
        public static let imageURLs = "imageURLs"
        public static let ownerID = "ownerID"
        /// Ключ партиционирования по городам — то же поле `city`, что у заведения.
        public static let city = "city"
    }

    // MARK: - reviews/{id}

    public enum ReviewDoc {
        public static let venueID = "venueID"
        public static let authorID = "authorID"
        public static let authorName = "authorName"
        public static let rating = "rating"
        public static let text = "text"
        public static let photoEmojis = "photoEmojis"
        public static let photos = "photos"
        public static let createdAt = "createdAt"
        public static let updatedAt = "updatedAt"
        public static let hostReply = "hostReply"
        public static let itemID = "itemID"
        public static let itemName = "itemName"
        public static let verifiedVisit = "verifiedVisit"
        /// Ключ партиционирования по городам — то же поле `city`, что у заведения.
        public static let city = "city"
    }

    /// Вложенная карта `reviews/{id}.hostReply`.
    public enum HostReplyField {
        public static let text = "text"
        public static let createdAt = "createdAt"
        public static let updatedAt = "updatedAt"
    }

    // MARK: - hosts/{uid}

    public enum HostDoc {
        public static let businessName = "businessName"
        public static let categoryRaw = "categoryRaw"
        public static let phone = "phone"
        public static let email = "email"
        public static let verification = "verification"
        // Реквизиты / расширенная информация о бизнесе. Имена полей совпадают
        // с именами свойств `HostProfile`.
        public static let legalForm = "legalForm"
        public static let legalName = "legalName"
        public static let inn = "inn"
        public static let registrationAddress = "registrationAddress"
        public static let website = "website"
        public static let about = "about"
    }

    // MARK: - categories/{slug}

    public enum CategoryDoc {
        public static let slug = "slug"
        public static let name = "name"
        public static let icon = "icon"
        public static let emoji = "emoji"
        public static let order = "order"
        public static let enabled = "enabled"
    }

    // MARK: - Вложенные объекты внутри venues/{id}

    /// Элемент массива `items` (блюдо/услуга).
    public enum ItemField {
        public static let id = "id"
        public static let name = "name"
        public static let emoji = "emoji"
        public static let kind = "kind"
        public static let imageURL = "imageURL"
    }

    /// Элемент массива `branches` (филиал).
    public enum BranchField {
        public static let id = "id"
        public static let address = "address"
        public static let latitude = "latitude"
        public static let longitude = "longitude"
        public static let phone = "phone"
    }

    /// Элемент массива `weekHours` (часы одного дня).
    public enum DayHoursField {
        public static let closed = "closed"
        public static let open = "open"
        public static let close = "close"
    }

    /// Элемент массива `pointsBands` (диапазон чек→баллы).
    public enum PointsBandField {
        public static let maxAmount = "maxAmount"
        public static let points = "points"
    }

    /// Элемент массива `pointsRewards` (награда за баллы).
    public enum PointsRewardField {
        public static let id = "id"
        public static let type = "type"
        public static let title = "title"
        public static let cost = "cost"
        public static let ratio = "ratio"
        public static let active = "active"
    }

    // MARK: - Кошельки гостя

    public enum CouponDoc {
        public static let code = "code"
        public static let userID = "userID"
        public static let venueID = "venueID"
        public static let venueName = "venueName"
        public static let title = "title"
        public static let kind = "kind"
        public static let dealID = "dealID"
        public static let used = "used"
        public static let createdAt = "createdAt"
    }

    public enum LoyaltyCardDoc {
        public static let userID = "userID"
        public static let venueID = "venueID"
        public static let venueName = "venueName"
        public static let stamps = "stamps"
        public static let completedRounds = "completedRounds"
        public static let goal = "goal"
        public static let reward = "reward"
    }

    public enum VenuePointsDoc {
        public static let userID = "userID"
        public static let venueID = "venueID"
        public static let venueName = "venueName"
        public static let balance = "balance"
        public static let lifetimeEarned = "lifetimeEarned"
        public static let lifetimeRedeemed = "lifetimeRedeemed"
    }

    public enum BonusGrantDoc {
        public static let userID = "userID"
        public static let amount = "amount"
        public static let claimed = "claimed"
    }

    /// `redemptions/{userID}_{dealID}` — детерминированный id, повторов не будет.
    public enum RedemptionDoc {
        public static let userID = "userID"
        public static let dealID = "dealID"
        public static let venueID = "venueID"
        public static let createdAt = "createdAt"
        public static let status = "status"
        /// Значение поля `status` для свежей записи (её досчитывает Cloud Function).
        public static let statusNew = "new"
    }

    /// `referrals/{inviteeID}` — один документ на приглашённого.
    public enum ReferralDoc {
        public static let inviteeID = "inviteeID"
        public static let referrerID = "referrerID"
        public static let createdAt = "createdAt"
    }

    /// `giftCoupons/{code}` — подарочный купон, забирается один раз.
    public enum GiftCouponDoc {
        public static let title = "title"
        public static let code = "code"
        public static let fromName = "fromName"
        public static let claimed = "claimed"
        public static let claimedAt = "claimedAt"
        public static let createdAt = "createdAt"
    }

    // MARK: - Телеметрия и рассылки

    /// `analyticsEvents/{auto}` — клиентское событие; счётчики досчитывает
    /// Cloud Function (писать в `analytics/*` напрямую клиенту запрещено).
    public enum AnalyticsEventDoc {
        public static let venueID = "venueID"
        public static let metric = "metric"
        public static let createdAt = "createdAt"
    }

    /// `analytics/{venueID}/days/{yyyy-MM-dd}` — счётчики за день.
    /// Все поля, кроме `date`, суммируются как метрики.
    public enum AnalyticsDayDoc {
        public static let date = "date"
    }

    /// `userTokens/{fcmToken}` — устройство для адресной рассылки.
    public enum UserTokenDoc {
        public static let city = "city"
        public static let uid = "uid"
        public static let updatedAt = "updatedAt"
    }

    /// `pushCampaigns/{auto}` — заявка на рассылку; админ одобряет в панели
    /// (`status` → `approved`), дальше рассылает Cloud Function.
    public enum PushCampaignDoc {
        public static let headline = "headline"
        public static let body = "body"
        public static let city = "city"
        public static let category = "category"
        public static let venueID = "venueID"
        public static let dealID = "dealID"
        public static let ownerID = "ownerID"
        public static let status = "status"
        public static let delivered = "delivered"
        public static let createdAt = "createdAt"
        /// Значение `status` у только что созданной заявки.
        public static let statusPending = "pending"
    }

    // MARK: - Ответы Cloud Functions (scanCoupon / redeemVenuePoints)
    //
    // Это не Firestore, а тело HTTP-ответа. Держим здесь по той же причине:
    // ключи заданы сервером в functions/src/index.ts и должны совпадать.

    public enum ScanResponse {
        public static let ok = "ok"
        public static let title = "title"
        public static let loyalty = "loyalty"
        public static let stamps = "stamps"
        public static let goal = "goal"
        public static let rewardIssued = "rewardIssued"
        public static let rewardTitle = "rewardTitle"
        public static let error = "error"
        public static let points = "points"
        public static let awarded = "awarded"
        public static let balance = "balance"
        // Поля запроса.
        public static let code = "code"
        public static let venueID = "venueID"
        public static let billAmount = "billAmount"
        public static let bandIndex = "bandIndex"
        public static let idempotencyKey = "idempotencyKey"
        /// true — сервер распознал повтор по ключу и НЕ начислил снова.
        public static let replayed = "replayed"
    }

    /// Data-payload push-уведомления (его собирает `sendPushCampaign`).
    /// По этим ключам приложение решает, что открыть по тапу.
    public enum PushPayload {
        public static let dealID = "dealID"
        public static let venueID = "venueID"
    }

    public enum RedeemResponse {
        public static let ok = "ok"
        public static let redeemed = "redeemed"
        public static let balance = "balance"
        public static let rewardTitle = "rewardTitle"
        public static let somOff = "somOff"
        public static let error = "error"
        /// true — сервер распознал повтор по idempotencyKey и НЕ списал баллы снова.
        public static let replayed = "replayed"
        public static let idempotencyKey = "idempotencyKey"
        // Поля запроса.
        public static let venueID = "venueID"
        public static let userID = "userID"
        public static let rewardId = "rewardId"
        public static let pointsToSpend = "pointsToSpend"
    }
}

// MARK: - Кодирование enum'ов в строковые ключи Firestore

/// Слаги категорий/типов/статусов — тоже часть схемы: их пишет и админка.
public enum FSKeys {
    public static let category: [String: VenueCategory] = [
        "cafe": .cafe, "coffee": .coffee, "fastfood": .fastfood,
        "restaurant": .restaurant, "teahouse": .teahouse, "bakery": .bakery,
    ]

    public static let dealType: [String: DealType] = [
        "discount": .discount, "promo": .promo, "novelty": .novelty, "announcement": .announcement,
    ]

    public static let dealStatus: [String: DealStatus] = [
        "active": .active, "paused": .paused, "expired": .expired, "draft": .draft,
    ]

    // Обратные карты: модель → строковый ключ для записи.
    private static let categoryKey: [VenueCategory: String] = [
        .cafe: "cafe", .coffee: "coffee", .fastfood: "fastfood",
        .restaurant: "restaurant", .teahouse: "teahouse", .bakery: "bakery",
    ]
    private static let dealTypeKey: [DealType: String] = [
        .discount: "discount", .promo: "promo", .novelty: "novelty", .announcement: "announcement",
    ]

    /// Пользовательские категории из админки пишем как есть (`rawValue`).
    public static func key(for c: VenueCategory) -> String { categoryKey[c] ?? c.rawValue }
    public static func key(for t: DealType) -> String { dealTypeKey[t] ?? t.rawValue }
}
